impeller/renderer/backend/gles/device_buffer_gles.cc

```diff
@@ -78,11 +85,13 @@ bool DeviceBufferGLES::BindAndUploadDataIfNecessary(BindingType type) const {
 
   gl.BindBuffer(target_type, buffer.value());
 
-  if (!uploaded_) {
+  if (dirty_) {
     TRACE_EVENT0("impeller", "BufferData");
-    gl.BufferData(target_type, data_->GetSize(), data_->GetMapping(),
+    gl.BufferData(target_type, buffer_->GetLength(), buffer_->GetBuffer(),
```

Comment:
I realize this is a bit yak-shavy, but there is a pessimization here where the entire arena will get uploaded even when only a sub-range has been modified.

In a later patch, we could setup the backing store to use a sparse mmap'ed arena and memmove into it any data the application provides. Then, note the range that must be flushed. During the upload, use glSubTexImage2D to upload the arenas that need to be flushed (after consolidation for overlapping ranges and such).

It is just the tedium of dealing with GLES 2.0 :/

A yak-shave within a yak-shave: impeller/base needs an mmaped allocation for sparse arenas. We just never ended up needing these in the Metal backend (and will probably not need them for anything other than ES). The posix stuff ought to be easy. The windows impl may need someone with a Windows box.

Reply:
Oh yes, for sure. I added a bug [here](https://github.com/flutter/flutter/issues/104447). Did you mean `glBufferSubData` (as opposed to `glSubTexImage2D`)? Although having sparse texture uploading would be good too.

Having the backing store be an mmapped arena sounds like a good idea -- just to make sure we're on the same page: The aim would be to make the arena have a per-upload lifetime (flush), that way we only ever hang onto enough memory to keep track of what needs uploading, and avoid needlessly mirroring all the device data.

Reply:
Right. The bug looks good.